src/Build/Construction/ProjectMetadataElement.cs

```diff
@@ -100,13 +100,17 @@ public string Value
         /// Creates an unparented ProjectMetadataElement, wrapping an unparented XmlElement.
         /// Caller should then ensure the element is added to a parent.
         /// </summary>
-        internal static ProjectMetadataElement CreateDisconnected(string name, ProjectRootElement containingProject)
+        internal static ProjectMetadataElement CreateDisconnected(string name, ProjectRootElement containingProject, ElementLocation location = null)
         {
             XmlUtilities.VerifyThrowArgumentValidElementName(name);
             ErrorUtilities.VerifyThrowArgument(!FileUtilities.ItemSpecModifiers.IsItemSpecModifier(name), "ItemSpecModifierCannotBeCustomMetadata", name);
             ErrorUtilities.VerifyThrowInvalidOperation(!XMakeElements.ReservedItemNames.Contains(name), "CannotModifyReservedItemMetadata", name);
 
             XmlElementWithLocation element = containingProject.CreateElement(name);
+            if (location != null)
+            {
+                element.Location = location;
```

Comment:
nit: I wonder if there's a slightly more elegant way of setting location for newly created elements than adding a Location setter. If you passed the location to `ProjectRootElement.CreateElement` which in turn would pass it to a new overload of  `XmlDocumentWithLocation.CreateElement`, you could set its `_reader` field for the duration of the base call, similar to what `Load` does. Note that `_reader` is of type `IXmlLineInfo`, so not necessarily an XML reader. Or you could introduce a new field (`_currentLocation`?) , which would then even avoid boxing.

Overall, I would find it nicer to leave the `*WithLocation` classes concerned with setting location without leaking the setter like this. What do you think?

Reply:
Thanks for looking into this! Great suggestion!

I was as well trying to see if I can just extend the creation and avoid subsequent mutation.
The problem that stopped me and let me to current code is that the `XmlDocumentWithLocation` extends `XmlDocument` and our overriden `CreateElement` is not being called directy, but via a `XmlDocument` function that inside calls internal helper function:

<img width="1015" alt="image" src="https://user-images.githubusercontent.com/3809076/231559925-0714d99b-b7aa-4fdf-b16a-9add2f899374.png">


So we'd need to duplicate some of that internal `XmlDocument` functionality (and maintain it going forward).

Not sure which is the lesser evil (allowing mutations vs duplicating `XmlDocument` code) - I arbitrarily decided for openng for mutation.

What do you think about this?

Reply:
Great thanks for offline discussion and tips! - I ended up passing the info via `AsyncLocal` filed.